### PR TITLE
Relaxed alignment constraints for memory and array accesses

### PR DESCRIPTION
**Pull request by vbgl**
Please contribute suggestions for concrete syntax.

### CODE REVIEW

**Review by bgregoir — DISMISSED** [review]
For me it is ok. Should we want more for merging it ?

**Comment by vbgl** [comment]
Let me check a few more things.

**Comment by vbgl** [comment]
There should be an annotation in slicing operations, at least for consistency.

File: compiler/src/toEC.ml [hunk 1]
```diff
@@ -1125,11 +1125,11 @@ module Leak = struct
 
   let safe_lval pd env = function
     | Lnone _ | Lvar _ -> []
-    | Lmem(ws, x, e) -> 
+    | Lmem(al, ws, x, e) -> (* FIXME: alignment *)
```
Comment (eponier) [inline compiler/src/toEC.ml:1128]: Here also.

**Review by eponier — APPROVED** [review]
A few comments that you can decide to ignore. I let you squash and/or merge depending on what you find the cleanest solution.

File: proofs/lang/memory_model.v [hunk 2]
```diff
@@ -318,20 +372,23 @@ Section CoreMem.
...
+    write m1 aw pw vw = ok m2 ->
+    write m1' aw pw vw = ok m2' ->
+    read m2 ar pr U8 = read m2' ar pr U8.
   Proof.
     move=> hr hw hw'.
     by rewrite (write_read8 hw) (write_read8 hw') /=; case: andP.
  Qed.
 
+ Definition disjoint_range_ovf p s p' s' : Prop :=
+   ∀ i i' : Z, 0 <= i < s → 0 <= i' < s' → add p i ≠ add p' i'.
```
[the remark this answers is not in the document]
Comment (vbgl) [reply proofs/lang/memory_model.v]: Done.

File: proofs/compiler/stack_alloc.v [hunk 3]
```diff
@@ -176,10 +176,10 @@ Record pos_map := {
...
 (* TODO: Z.land or is_align ?
    Could be just is_align (sub_region_addr sr) ws ? *)
-Definition check_align x (sr:sub_region) ws :=
-  Let _ := assert (ws <= sr.(sr_region).(r_align))%CMP
+Definition check_align al x (sr:sub_region) ws :=
+  Let _ := assert ((al == Unaligned) || (ws <= sr.(sr_region).(r_align))%CMP) (* FIXME: is this check needed? *)
                   (stk_ierror_basic x "unaligned offset") in
-  assert (Z.land sr.(sr_zone).(z_ofs) (wsize_size ws - 1) == 0)%Z
+  assert ((al == Unaligned) || (Z.land sr.(sr_zone).(z_ofs) (wsize_size ws - 1) == 0)%Z)
          (stk_ierror_basic x "unaligned sub offset").
```
[the remark this answers is not in the document]
Comment (vbgl) [reply proofs/compiler/stack_alloc.v]: What is `sub_region_addr`?

Turned the FIXME into a TODO (for a stack-alloc expert).

File: proofs/compiler/stack_alloc_proof.v [hunk 4]
```diff
@@ -927,15 +920,15 @@ Section EXPR.
...
     (forall zofs, ofs = Some zofs -> 0 <= zofs /\ zofs + wsize_size ws <= size_slot x.(gv)) ->
     get_var_kind pmap x = ok (Some vpk) ->
-    check_vpk_word rmap x.(gv) vpk ofs ws = ok t ->
+    check_vpk_word rmap al x.(gv) vpk ofs ws = ok t ->
     exists sr bytes, [/\
       check_gvalid rmap x = Some (sr, bytes),
       let isub_ofs := interval_of_zone (sub_zone_at_ofs sr.(sr_zone) ofs (wsize_size ws)) in
       ByteSet.mem bytes isub_ofs &
-      is_align (sub_region_addr sr) ws].
+      is_aligned_if al (sub_region_addr sr) ws].
```
[the remark this answers is not in the document]
Comment (vbgl) [reply proofs/compiler/stack_alloc_proof.v:931]: No way.

File: compiler/src/printFexpr.ml [hunk 5]
```diff
@@ -13,9 +13,9 @@ let rec pp_fexpr fmt =
...
   function
   | Rexpr e -> pp_fexpr fmt e
-  | Load (sz, x, e) -> Format.fprintf fmt "(%a)[%a + %a]" pp_wsize sz pp_var_i x pp_fexpr e
+  | Load (al, sz, x, e) -> Format.fprintf fmt "(u%a)[%a + %a%a]" pp_wsize sz pp_var_i x pp_fexpr e pp_aligned al
 
 let pp_lexpr fmt =
   function
   | LLvar x -> pp_var_i fmt x
-  | Store (sz, x, e) -> Format.fprintf fmt "(%a)[%a + %a]" pp_wsize sz pp_var_i x pp_fexpr e
+  | Store (al, sz, x, e) -> Format.fprintf fmt "(u%a)[%a + %a%a]" pp_wsize sz pp_var_i x pp_fexpr e pp_aligned al
```
[the remark this answers is not in the document]
Comment (vbgl) [reply compiler/src/printFexpr.ml]: Well spotted. Fixed.

File: compiler/src/printCommon.ml [hunk 6]
```diff
@@ -14,6 +14,15 @@ let pp_wsize fmt sz = fprintf fmt "%a" pp_string0 (string_of_wsize sz)
 
 (* -------------------------------------------------------------------- *)
 
+let pp_aligned fmt =
+  function
+  | Memory_model.Aligned ->
+     Format.fprintf fmt "#aligned "
+  | Unaligned ->
+     Format.fprintf fmt "#unaligned "
+
```
[the remark this answers is not in the document]
Comment (vbgl) [reply compiler/src/printCommon.ml:23]: I’ve tried to clean this up.

File: compiler/src/pretyping.ml [hunk 7]
```diff
@@ -950,18 +958,19 @@ let rec tt_expr pd ?(mode=`AllVar) (env : 'asm Env.env) pe =
...
     let ws = Option.map_default tt_ws (P.ws_of_ty ty) ws in
     let ty = P.tu ws in
     let i,ity  = tt_expr ~mode pd env pi in
     let i = ensure_int (L.loc pi) i ity in
     begin match olen with
-    | None -> P.Pget (aa, ws, x, i), ty
+    | None -> P.Pget (al, aa, ws, x, i), ty
     | Some plen ->
       let len,ity  = tt_expr ~mode:`OnlyParam pd env plen in
       check_ty_eq ~loc:(L.loc plen) ~from:ity ~to_:P.tint;
```
[the remark this answers is not in the document]
Comment (vbgl) [reply compiler/src/pretyping.ml]: Added a warning.

File: compiler/src/syntax.ml [hunk 8]
```diff
@@ -163,7 +163,7 @@ type pexpr_r =
 
 and pexpr = pexpr_r L.located
 
-and mem_access = wsize option * pident * ([`Add | `Sub] * pexpr) option
+and mem_access = wsize option * pident * ([`Add | `Sub] * pexpr) option * [`Aligned|`Unaligned] option
```
[the remark this answers is not in the document]
Comment (vbgl) [reply compiler/src/syntax.ml]: Done.

File: compiler/src/toEC.ml [hunk 9]
```diff
@@ -1085,17 +1085,17 @@ module Leak = struct
         | Arr(ws,n) -> Inita (L.unloc x, arr_size ws n) :: safe
         | _ -> Initv(L.unloc x) :: safe 
       else safe 
-    | Pload (ws,x,e) -> 
+    | Pload (al, ws,x,e) -> (* FIXME: alignment *)
```
[the remark this answers is not in the document]
Comment (vbgl) [reply compiler/src/toEC.ml]: Changed into a TODO: this whole extration-for-safety is to be rewritten.

File: compiler/src/toEC.ml (hunk 1 above)
Comment (vbgl) [reply compiler/src/toEC.ml]: Same.

**Comment by vbgl** [comment]
Thanks Jean-Christophe for your careful review. I think I’ve addressed your concerns.

File: proofs/compiler/stack_alloc.v (hunk 3 above)
[the remark this answers is not in the document]
Comment (eponier) [reply proofs/compiler/stack_alloc.v]: I'll have to change that function when implementing subarray with unknown index anyway.

File: proofs/compiler/stack_alloc_proof.v (hunk 4 above)
[the remark this answers is not in the document]
Comment (eponier) [reply proofs/compiler/stack_alloc_proof.v:931]: Fair enough.

File: compiler/src/pretyping.ml (hunk 7 above)
[the remark this answers is not in the document]
Comment (eponier) [reply compiler/src/pretyping.ml]: Nice! But the same change should be done for PLArray. And in the same spirit, we should not print `#aligned` for subarrays as is currently done.

**Comment by eponier** [comment]
The opam job was killed in the CI, similarly to https://github.com/jasmin-lang/jasmin/pull/560#issuecomment-2014845825. Do you have an idea of what happens? Maybe too much RAM used by the job? We should fix that in the long run, not for this PR of course.

File: compiler/src/pretyping.ml (hunk 7 above)
[the remark this answers is not in the document]
Comment (vbgl) [reply compiler/src/pretyping.ml]: Done.

**Review by eponier — APPROVED** [review]

File: compiler/src/pretyping.ml (hunk 7 above)
[the remark this answers is not in the document]
Comment (vbgl) [reply compiler/src/pretyping.ml]: I think I had misunderstood this remark. Here is a fix: #772

[the remark this answers is not in the document]
Comment (eponier) [reply compiler/src/pretyping.ml]: I was indeed thinking about printing in Jasmin. Thanks for the fix!

**Pull request merged**